### ISSUE TITLE
rpcserver: panic

### ISSUE THREAD

**Issue by dajohi** [btcsuite/btcd#846]
If I loop through disconnecting all peers by ID, I occassionally see:

```
2016/11/07 20:16:51 http: panic serving [::1]:27948: runtime error: invalid memory address or nil pointer dereference                            [377/91152]
goroutine 1457826 [running]:
net/http.(*conn).serve.func1(0xc423a4b800)
        /usr/local/go/src/net/http/server.go:1491 +0x12a
panic(0x9a3320, 0xc4200100d0)
        /usr/local/go/src/runtime/panic.go:458 +0x243
main.handleGetPeerInfo(0xc4204d6d20, 0x953f20, 0xe5c968, 0xc429e08420, 0xc42006a980, 0xc428ad1a01, 0xc428ad1ab0, 0x430a6c)
        /home/dhill/.go/src/github.com/btcsuite/btcd/rpcserver.go:2216 +0x29e
main.(*rpcServer).standardCmdResult(0xc4204d6d20, 0xc4257b5b80, 0xc429e08420, 0xc42e37ea80, 0xc429e08420, 0x0, 0x0)
        /home/dhill/.go/src/github.com/btcsuite/btcd/rpcserver.go:3911 +0x97
main.(*rpcServer).jsonRPCRead(0xc4204d6d20, 0xe144a0, 0xc44c473790, 0xc4270c0c30, 0x1)
        /home/dhill/.go/src/github.com/btcsuite/btcd/rpcserver.go:4067 +0xa0b
main.(*rpcServer).Start.func1(0xe144a0, 0xc44c473790, 0xc4270c0c30)
        /home/dhill/.go/src/github.com/btcsuite/btcd/rpcserver.go:4136 +0x1da
net/http.HandlerFunc.ServeHTTP(0xc42093c7f0, 0xe144a0, 0xc44c473790, 0xc4270c0c30)
        /usr/local/go/src/net/http/server.go:1726 +0x44
net/http.(*ServeMux).ServeHTTP(0xc42067e8d0, 0xe144a0, 0xc44c473790, 0xc4270c0c30)
        /usr/local/go/src/net/http/server.go:2022 +0x7f
net/http.serverHandler.ServeHTTP(0xc4200a8280, 0xe144a0, 0xc44c473790, 0xc4270c0c30)
        /usr/local/go/src/net/http/server.go:2202 +0x7d
net/http.(*conn).serve(0xc423a4b800, 0xe14f20, 0xc4257b4f40)
        /usr/local/go/src/net/http/server.go:1579 +0x4b7
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:2293 +0x44d
```

**Comment by davecgh** [comment]
Hopefully @tuxcanfly can submit a PR to fix.  I suspect the issue is that the `conn` field is not set yet due to `peer.Connected` and `peer.AssociateConnection` are only using a single atomic at the top instead of protecting the entire thing with a mutex.  The `handleGetPeers` func calls `server.Peers` which in turn loops the peers and calls `.Connected` which checks the atomic, but looking at `peer.AssociateConnection` the atomic is set first, so it can race.

**Comment by tuxcanfly** [comment]
I was able to repro this by using `time.Sleep` in `handleGetPeerinfo` after `Peers()` before `GetPeerInfoResult`.

I can confirm the panic was being caused by a dangling disconnected peer, which is disconnecting during the `getpeerinfo` rpc call.  I have proposed a fix in #853 to make `LocalAddr` continue returning local address after the peer disconnects.

I don't think the race condition mentioned above is possible, at least with respect to this issue, because `p.Connected` was being called only after `p.AssociateConnection` is called.
